python/adelphi/adelphi/anonymize.py

```diff
@@ -38,59 +33,70 @@ def get_name(original_name, prefix):
-    udt.field_names = [get_name(field_name, FIELD_PREFIX) for field_name in udt.field_names]
-    # field types
-    udt.field_types = [get_name(field_type, TYPE_PREFIX)
-                       if field_type in name_map[TYPE_PREFIX]
-                       else field_type
-                       for field_type in udt.field_types]
 
+def anonymize_udts(udts):
+    # anonymize all udt names first
+    for udt in udts:
+        udt.keyspace = get_name(udt.keyspace, KEYSPACE_PREFIX)
+        udt.name = get_name(udt.name, TYPE_PREFIX)
+    
+    for udt in udts:
+        # field names
+        udt.field_names = [get_name(field_name, FIELD_PREFIX) for field_name in udt.field_names]
+        # field types
+        udt.field_types = [anonymize_type(field_type) for field_type in udt.field_types]
 
-def anonymize_column(column, prefix):
-    column.name = get_name(column.name, prefix)
+def anonymize_column(column):
+    column.name = get_name(column.name, COLUMN_PREFIX)
+    column.cql_type = anonymize_type(column.cql_type)
 
 
 def anonymize_index(index):
     index.name = get_name(index.name, INDEX_PREFIX)
-    prefix = COLUMN_PREFIX if index.index_options['target'] in name_map[COLUMN_PREFIX] \
-        else CLUSTERING_KEY_PREFIX
-    index.index_options['target'] = name_map[prefix][index.index_options["target"]]
-    index.keyspace_name = name_map[KEYSPACE_PREFIX][index.keyspace_name]
-    index.table_name = name_map[TABLE_PREFIX][index.table_name]
-
+    index.index_options['target'] = get_name(index.index_options["target"], COLUMN_PREFIX)
+    index.keyspace_name = get_name(index.keyspace_name, KEYSPACE_PREFIX)
+    index.table_name = get_name(index.table_name, TABLE_PREFIX)
 
 def anonymize_table(table):
     table.keyspace_name = get_name(table.keyspace_name, KEYSPACE_PREFIX)
     table.name = get_name(table.name, TABLE_PREFIX)
 
-    for partition_key in table.partition_key:
-        anonymize_column(partition_key, PARTITION_KEY_PREFIX)
+    for pk in table.partition_key:
+        anonymize_column(pk)
 
-    for clustering_key in table.clustering_key:
-        anonymize_column(clustering_key, CLUSTERING_KEY_PREFIX)
+    for ck in table.clustering_key:
+        anonymize_column(ck)
 
-    # CK are also in the standard columns, but different objects
-    # if we don't anonymize them there too, the generated cql is wrong
-    for clustering_key in [t for t in table.columns.values() if t.name in name_map[CLUSTERING_KEY_PREFIX]]:
-        clustering_key.name = name_map[CLUSTERING_KEY_PREFIX][clustering_key.name]
+    # remove comment since it may contain sensitive information
+    table.options["comment"] = ""
 
+    # The CQL python driver holds the same object references for clustering keys
+    # and regular columns, so we have to separate them otherwise the names
+    # will be re-anonymized.
     for column in get_standard_columns_from_table_metadata(table):
-        anonymize_column(column, COLUMN_PREFIX)
+        anonymize_column(column)
```

Comment:
Since we've now eliminated the distinction between clustering and partition keys and regular columns in an anonymization context can't we just simplify this logic to something like:

```python
for column in table.columns:
    anonymize_column(column)
```

Reply:
> can't we just simplify this logic to something like

not really, because the python driver references the same objects in `table.columns` and `table.clustering_key`, so some columns would go through the anonymization twice. I added a comment in the code trying to clarify that.

Maybe we could skip the anonymization for `table.clustering_key` all together and simply rely on the `table.columns`, but if something changes internally in a future version of the driver and they don't have the same references anymore, we'll fail to anonymize some names. It seemed this was the most future-proof alternative.

Reply:
Yeah, that second option is what I was after; don't do anything for clustering keys and just iterate through columns.  I agree that future changes to the driver could possibly break that but the only assumption we're making there is that table.columns gives us an iterator for every column in the table (regardless of it's role).  Breaking _that_ contract would be a significant change for a large number of users of the Python driver... so it seems like we're on pretty safe ground there.

Plus, I mean... it's not like there are a slate of impending Python driver changes going forward. :)

_And_ we're looking at pinning the driver version with the compatible operator, and since any future driver version that _did_ make the change mentioned above would _at least_ need a new minor version (more correctly a new major version for something so drastic) we'd be well insulated from future change.

So I guess I'd argue we're on pretty safe ground here. :)

Reply:
> And we're looking at pinning the driver version with the compatible operator

your PR that got merged recently didn't pin to the exact version, right? it gives some flexibility to upgrade to a newer patch version right?

Anyway, I'm leaning towards leaving it the way it is since it reduces the risk of leakage to zero and removing it doesn't buy us much.

Reply:
Yeah, the PR uses the "compatible" operator which amounts to allowing different patch versions with the same major/minor.  It's true that this does allow the version to change, but you'll only have changes in the patch version... and by definition a patch change won't introduce API changes.  So I guess I see the risk here as _extremely_ low.

Broadly speaking I agree, it isn't a _huge_ thing... I'm mostly interested in trying to simplify the code as much as possible.  Looking at this code six months from now one could very credibly wonder why we're still looping over these various sets of columns instead of... just looping over the columns.  It's not gonna make or break anything, it's just simpler.

Reply:
> and by definition a patch change won't introduce API changes

Since we're mutating internal state of the Metadata object (which in general is supposed to be read-only), I'm not really sure if that qualifies as an API. I think that's mostly where my concern is.

Reply:
Yeah, I'm... not really thrilled with that, to put it mildly.  FWIW the second Github PR sought to address this by changing adelphi.store.with_cluster to call the passed cluster_fn with a cluster connection, close that connection and return the object.  So even though we're dealing with a Metadata instance we're not retaining a conn to the cluster anymore... so there's no obvious way for that data to be persisted.

We _are_ still modifying the Metadata object itself locally, which I agree isn't great... I really wish that object were immutable.  But that's a different concern from what we were discussing originally (i.e. column traversal).

We can leave it as-is and maybe consider taking a look at it later.  As I think we've both mentioned above it isn't by any means essential that we change this now; we can always revisit.

Reply:
Right, we grab the Metadata and close the connection, that's how it works from the beginning.

In my comment I wasn't concerned about the immutability nor the data being persistent, what I was trying to point at was the fact that by changing the internal state directly it doesn't respect any API contract, in other words, there's no guarantee the internal object references won't change even in a patch release, which would potentially break the column traversal.

Anyway, I still think the leakage of data is too big of a problem to run any risk around it, I prefer to have the anonymization completely under our control.

Reply:
I still disagree, but not strongly enough to hold up the PR further... and it's something of a speculative concern anyway.  We can revisit in the future if needed.

FWIW my initial comment above (re: going back to get connection then returning Metadata) referred to an explicit change that happened in the first Github PR.  At some point prior to that (I forget exactly when) I became increasingly uncomfortable with the returned Metadata instance and changed the relevant logic.  Everything (evaluating the Metadata object, building up state, and printing the output) now happened within an input cluster_fn which took the Cluster instance as an arg; there was no return statement out of that logic.

Then I got nervous about setting RF on a Metadata object while still connected to the cluster and changed it back.  That's what you see at https://github.com/datastax/adelphi/pull/51/files#diff-03a41fd89b9e4d86cb31cefbd02da1b19aed39fb8dc9cbaea15f80b672c25d42

I mention this only to say that it wasn't _always_ the case that we return a Metadata object from the equivalent of store.with_cluster.  _That's_ why I went into the detail in my original comment.